New functions

PiotrWeppo opened the pull request:
Pull request z grubsza skupia się wokół czterech rzeczy:

- Do menu budowy wiadomości została dodana opcja usuwania wybranego pola tekstowego. Przedtem można było jedynie dodawać pola tekstowe, a jedynym sposobem pozbycia się ich, było zresetowanie wszystkich zmian w wiadomości.
- Dodano nową komendę /embed_update, która pozwala edytować ostatnio wysłaną wiadomość.
- Ostatnio wysłana wiadomość automatycznie aktualizuje główne pole tekstowe co 15 sekund.
- Posprzątanie kodu i części błędów.

Issue: https://github.com/KNR-PW/discord-bot/issues/8

PiotrWeppo committed:
Added count_members, list_members functions.

PiotrWeppo committed:
Extended list_members and count_members functions to allow starting messages with "not" operators, and also searching for roles using only "not" operators

PiotrWeppo committed:
Work in progress

PiotrWeppo committed:
Added function that parses list_members function inside the code.

PiotrWeppo committed:
Fixed issues.
Added count_members command to the parsing function.
Parsed message can be read in the embed when typed !embed in discord chat.

PiotrWeppo committed:
Added functionality to show specific role in parsed text.

PiotrWeppo committed:
Added function to parse discord username into text.

PiotrWeppo committed:
Code cleanup

PiotrWeppo committed:
Packed code into smaller functions for better readability.

PiotrWeppo committed:
Added docstrings to functions. Changed some variables' and functions' names for better readability.

PiotrWeppo committed:
Added type hints.
Further code cleanup.

PiotrWeppo committed:
Updated README file.
Got rid of linting error.

PiotrWeppo committed:
Minor update to README

PiotrWeppo committed:
Added searching for voice and text channels to the list of possible commnads.
README update.

PiotrWeppo committed:
Added new classes to dynamically generate embed message from discord embed.

PiotrWeppo committed:
Modified README file.
Added gif showing bot usage.

PiotrWeppo committed:
Added help command to the bot.
Changes in readme.

PiotrWeppo committed:
Modified the function that looks for commands to be converted in the text.
The function now ignores spaces at the beginning and end of the text.

PiotrWeppo committed:
Merge remote-tracking branch 'upstream/main' into new_functions

PiotrWeppo committed:
Added option to remove  message fields from the embed.

PiotrWeppo committed:
Added slash command for updating last sent embed.

PiotrWeppo committed:
linters' warnings fixes,
markdown file edit

PiotrWeppo committed:
A newly sent message automatically updates its description at the specified time.

PiotrWeppo committed:
Bug fixes and code cleanup

pktiuk commented:
W pull requeście powinno się przedstawić jakiś kontekst. Tytuł `New Functions` zupełnie nic nie mówi osobie, która bedzie potencjalnie przeglądać potem historię tego repozytorium. Poza tym w opisie warto byłoby także zawrzeć jakiś szerszy kontekst. Napisać co się chce osiągnąć/zaimplementować, i w związku z jakim zadaniem(issue).   
Proszę o wyedytowanie tego Pull requesta.

pktiuk commented on bot.py:
```diff
@@ -3,13 +3,23 @@
 load_dotenv()  # loads your local .env file with the discord token
+DISCORD_TOKEN: Optional[str] = os.getenv("DISCORD_TOKEN")
+
+# pylint: disable=invalid-name
+embed_channel_id = int()
+embed_message_id = int()
+embed_description = str()
+# pylint: enable=invalid-name
```
Czy dobrze rozumiem, że jeśli wyłączymy i włączymy bota to utraci on wiedzę o tym jak znaleźć swoją ostatnią wiadomość?

PiotrWeppo replied:
Tak, bot nie korzysta w tym momencie z zewnątrznej bazy danych.

PiotrWeppo replied:
Czy chcesz, żeby było coś w rodzaju zapisu do pliku tekstowego w folderze bota? Mogę również dodać coś w rodzaju formularza, gdzie w przypadku zresetowania bota, można wkleić i  przypomnieć ID ostatniej wiadomości. W przypadku drugiej opcji pojawia się niestety problem potrzeby manualnego przypominania oraz brak możliwości odświeżania wiadomości, przed jej przypomnieniem.

pktiuk replied:
Zapisywanie tego typu rzeczy konfiguracyjnych w jakimś pliku byłoby bardzo wygodne i łatwe do zimplementowania.   

Ale najlepiej będzie użyć do tego gotowej biblioteki.   
Spójrz tutaj: https://docs.python.org/3/library/configparser.html

PiotrWeppo replied:
Zrobiłem tak jak zasugerowałeś.

pktiuk commented on bot.py:
```diff
@@ -273,6 +327,50 @@ async def add_field(self, interaction: discord.Interaction) -> None:
+    """
+    Subclass of the `discord.ui.View` class. Used for creating a select prompt.
+
+    Args:
+        placeholder (str): The placeholder text that will be displayed
+        in the channel select menu.
+        options (List[SelectOption]): A list of `SelectOption` instances that
+        will be displayed as options in the select prompt.
+        max_values (int, optional): The maximum number of options
+        that can be selected by the user. Default is 1.
+        ephemeral (bool, optional): A boolean indicating whether
+        the select prompt will be sent as an ephemeral message or not. Default is False.
+    """
+
+    def __init__(
+        self,
+        placeholder: str,
```
Czy opis argumentów konstruktora nie powinien się znajdować w ciele konstruktora, a nie ciele klasy?

PiotrWeppo replied:
To skomplikowane. Na podstawie mojej wiedzy, opis argumentów powinien znajdować się w ciele klasy. Odnośnie docstringów konstruktorów, z jednej strony znalazłem informacje, że powinny je również posiadać, z drugiej Pylint informuje mnie jedynie o braku docstringów ciał klas i ich metod, sugerując, że konstruktor nie potrzebuje takowego.

Odnośnik do tematu na stackoverflow:
https://stackoverflow.com/questions/62444612/should-i-list-class-methods-in-the-class-docstring

Przykład docstringów w stylu Google, z którego korzystam:
https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html

pktiuk replied:
```
The __init__ method may be documented in either the class level
    docstring, or as a docstring on the __init__ method itself.

    Either form is acceptable, but the two should not be mixed. Choose one
    convention to document the __init__ method and be consistent with it.
```
Oki, czyli wg konwencji obie opcje są dobre, tylko ważne, aby nie mieszać.  (osobiście nie do końca się zgadzam z tą konwencją, ale to tylko moja prywatna preferencja)

pktiuk commented on bot.py:
```diff
@@ -663,39 +854,21 @@ async def bot_help(ctx: commands.Context):
-@bot.command()
-async def hello(ctx):
-    """Reads the message from the chat. Returns the corresponding message.
+def finding_single_member(ctx, member_name: str) -> str:
+    """Takes the string and returns the corresponding member from the discord server.
```
Zmieniłbym nazewnictwo tej funkcji. na ogół nie używa się czasowników z końcówką -ing, bo to sugeruje, że to czynność trwająca, albo rzeczownik. w wolnym tłumaczeniu zamiast funkcji znalezienie członka powinniśmy mieć funkcję znajdź członka

PiotrWeppo replied:
Tak zrobię w kolejnym commicie i sprawdzę też nazwy innych funkcji

PiotrWeppo commented:
> W pull requeście powinno się przedstawić jakiś kontekst. Tytuł `New Functions` zupełnie nic nie mówi osobie, która bedzie potencjalnie przeglądać potem historię tego repozytorium. Poza tym w opisie warto byłoby także zawrzeć jakiś szerszy kontekst. Napisać co się chce osiągnąć/zaimplementować, i w związku z jakim zadaniem(issue). Proszę o wyedytowanie tego Pull requesta.

Jaka jest praktyka, w przypadku chęci wysłania wielu commitów, które dotykają róznych problemów. Czy powinienem otwierać jednego brancha do funkcjonalności z odpowiednią nazwą, drugiego do usuwania błędów itp.? Jak to się ma do przypadku, gdy robię kilka commitów zanim zrobię pull requesta?

pktiuk commented:
> Czy powinienem otwierać jednego brancha do funkcjonalności z odpowiednią nazwą, drugiego do usuwania błędów itp.?

Tak, wtedy można na szybko zmergować tego, który usuwa błędy (aby reszta zespołu miała naprawiony kod), a my tymczasem możemy dokładnie obgadać implementację nowych funkcjonalności. Dekompozycja (czyli rozdzielanie pracy na mniejsze kawałki) jest ważnym aspektem prac. Bo łatwiej jest zrobić 10 review po 100 lini kodu, niż jedno na 1000.

Oczywiście nie zawsze da się podielić tak pracę na drobne kawałki Ale jeśli się nie da to należy po prostu w opisie tak z grubsza opisać mniej więcej jakie rzeczy chcemy zrobić.

> Jak to się ma do przypadku, gdy robię kilka commitów zanim zrobię pull requesta?

Warto aby commit w jakiś sposób odzwierciedlały to co właśnie robisz, ale bez zbędnego rozdrabniania się, więc je też warto mieć uporządkowane, dlatego nie zaleca się wstawiać commitów typu "fix", albo "code cleanup", jeśli w ramach tego samego pull requesta stworzyłeś kod, który właśnie poprawiasz. Wtedy zaleca się używanie komendy `git rebase -i` [artykuł od gitlaba na ten temat](https://about.gitlab.com/blog/2020/11/23/keep-git-history-clean-with-interactive-rebase/)

PiotrWeppo commented:
> W pull requeście powinno się przedstawić jakiś kontekst. Tytuł `New Functions` zupełnie nic nie mówi osobie, która bedzie potencjalnie przeglądać potem historię tego repozytorium. Poza tym w opisie warto byłoby także zawrzeć jakiś szerszy kontekst. Napisać co się chce osiągnąć/zaimplementować, i w związku z jakim zadaniem(issue). Proszę o wyedytowanie tego Pull requesta.

Nie jestem pewien w jaki sposób mogę powiązać to z Issue

pktiuk commented:
PO prostu wklejasz linka do issue, a wtedy na stronie issue pojawia się link, że było ono omawiane w takim i takim PR-rze. Dzięki takiemu dwustronnemu wskazaniu można łatwo nawigować po większych repozytoriach i monitorować postępy.

PiotrWeppo committed:
Code shortening and minor fixes.

PiotrWeppo committed:
Added configparser module.
Now last sent message can be recalled even if bot was restarted.
Message information is dynamically written to and read from the configuration file.

PiotrWeppo committed:
Made python module that creates config.ini file.
If config.ini file is not found in the root folder it will be dynamically created during bot initialization.

PiotrWeppo committed:
Small fixes in bot.py and README.md

PiotrWeppo commented:
> > Czy powinienem otwierać jednego brancha do funkcjonalności z odpowiednią nazwą, drugiego do usuwania błędów itp.?
> 
> Tak, wtedy można na szybko zmergować tego, który usuwa błędy (aby reszta zespołu miała naprawiony kod), a my tymczasem możemy dokładnie obgadać implementację nowych funkcjonalności. Dekompozycja (czyli rozdzielanie pracy na mniejsze kawałki) jest ważnym aspektem prac. Bo łatwiej jest zrobić 10 review po 100 lini kodu, niż jedno na 1000.
> 
> Oczywiście nie zawsze da się podielić tak pracę na drobne kawałki Ale jeśli się nie da to należy po prostu w opisie tak z grubsza opisać mniej więcej jakie rzeczy chcemy zrobić.

To ma sens gdybym miał dostęp do mergowania własnego codu, albo robiłoby się to np. tego samego dnia. Jeżeli któregoś dnia mam dużo czasu i chcę rozwiązać kilka problemów na raz to nie mogę czekać po kilka dni, przed otworzeniem kolejnego brancha na kontynuację pracy.

PiotrWeppo committed:
Added two functions to configcreator.py to simplify code.

PiotrWeppo committed:
Merge pull request #2 from PiotrWeppo/new_functions

New functions

pktiuk commented on bot.py:
```diff
@@ -35,17 +38,91 @@ async def on_ready(self):
-    async def on_command_error(self, ctx, error):
-        await ctx.reply(error, ephemeral=True)
 
+CONFIG_FILE_EXISTS = os.path.exists("config.ini")
+if not CONFIG_FILE_EXISTS:
+    create_config_file()
```
Tą część bym przeniósł do funkcji czytających i piszących configi.   
W ten sposób cała logika związana z konfiguracją będzie w jednym pliku.

PiotrWeppo replied:
Zgoda

PiotrWeppo replied:
Zrobione

pktiuk commented on .gitignore:
```diff
@@ -130,3 +130,6 @@ dmypy.json
 .pyre/
+
+# config files
+.ini
```
```suggestion
# config files
*.ini
```
Tutaj jest potrzebny regex aby łapał wszystkie pliki z końcówką .ini

PiotrWeppo replied:
Zrobione

pktiuk commented on bot.py:
```diff
@@ -348,66 +486,63 @@ class EditSelectMenu(discord.ui.Select):
+                    label="Title and Message",
+                    description="Set a title and description for the embed",
+                ),
+                discord.SelectOption(
+                    label="Add Field", description="Add a field to the embed"
+                ),
+                discord.SelectOption(
+                    label="Remove Field", description="Remove a field from the embed"
+                ),
```
Wydaje mi się, że aby nie przytłoczyć użytkownika nie lepiej by było Zamiast AddField i RemoveField dać po prostu jedną opcję. Field i działałaby ona tak, `Title and Message`. 
Czyli jeśli jakieś wartości były już wpisane to po prostu pojawi się już wypełniony formularz.
To by bardzo ułatwiło testowanie widoku w podglądzie wiadomości (no i łatwe poprawianie błędów)
![image](https://github.com/KNR-PW/discord-bot/assets/45544416/141f91f0-c26d-4fac-89fd-0b6c67db78d0)

pktiuk replied:
Poza tym wydaje mi się, że dodawanie wielu pól tak jak na widocznym rysunku chyba nie jest potrzebna.   
Funkcja jest fajna, ale na ten moment chyba byłoby to trochę przekombinowane dla osób, które mają z tego docelowo korzystać

pktiuk replied:
Chociaż z drugiej strony trochę żal to usuwać.

Dałoby się po prostu dać `Edit Field: Lista osób` dla każdego z już utworzonych pól?
Aby dało się poprawić już istniejące pole bez pisania wszystkiego jeszcze raz?

PiotrWeppo replied:
Myślę, że tak. Tylko trzeba wziąć pod uwagę, że jest możliwość dodania do 25 pół tekstowych. Czy chciałbyś w takim razie, żeby menu tekstowe, które może mieć do 25 pól wyboru było dynamiczne i wraz z dodaniem pola, zwiększało się o wiersz z edycją wybranego pola? W jaki sposób pola mają być usuwane? Polem wyboru, które jest teraz czy czymś na zasadzie, że jeżeli usunie się treść pola i wyśle je, to znika?

pktiuk replied:
Tutaj niczego nie będę ci narzucać. Wszystko zależy od tego jak wtoim zdaniem będzie wygodniej i prościej w kodzie.

Priorytetem jest to, aby możliwe było łatwe edytowanie już wpisanej wiadomości, aby nie trzeba było wszystkiego wpisywać od zera w razie literówki.    
Możliwość posiadania wielu pól jest opcjonalna. Fajnie ją mieć, ale jeśli uznasz, że kod będzie prostszy to możesz bez żalu wyrzucić.

PiotrWeppo replied:
Jeżeli chodzi o kwestionariusze, które się wypełnia, to sprawdziłem, że maksymalnie może być 5 pól na raz, dlatego przy więcej niż jednym dodatkowym polu tekstu, trzeba by wprowadzić coś typu lista wyboru pola, która prowadzi do edycji zawartości konkretnego pola, ale wtedy gdyby się usunęło np. jakieś środkowe pole, a potem chciało dodać nowe na któreś miejsce, to musiałaby być kolejna lista wyboru, pytająca, gdzie ma być umieszczone nowe pole.
Daj znać jak to widzisz i czy chciałbyś coś takiego, czy zrezygnować z tego.

pktiuk replied:
To już byłoby chyba zbyt przekombinowane.

Chyba prościej jest zrezygnować z tego, bo te dodatkowe pola i tak są tylko dodatkiem, a nie kluczową funkcją.   
Pozostawiam ci wolną rękę e tym jak to zrobisz. W sumie jeśli chcesz to możesz to zostawić jak jest i w razie jak uzytkownicy docelowi będą narzekać to wtedy to się poprawi.

PiotrWeppo replied:
W takim razie na ten moment rezygnuję z rozbudowy tej funkcji.

PiotrWeppo replied:
W kolejnym branchu dodałbym zabezpieczenia, żeby tylko odpowiednie osoby mogły korzystać z komend.

PiotrWeppo replied:
+ zapomniałem jeszcze dodać śledzenie zmian w dodatkowych polach. Myślę, że ograniczę ich ilość do 5 np.

PiotrWeppo replied:
> W kolejnym branchu dodałbym zabezpieczenia, żeby tylko odpowiednie osoby mogły korzystać z komend.

Będę miał pytanie jaki chciałbyś żeby parametr był sprawdzany przy próbie wywołania komendy. W tym momencie tylko administrator serwera może używać komend, ale jest to bardzo "customizowalne". Począwszy od sprawdzania, czy wywołujący posiada jakąś specyficzną rolę, do sprawdzania, czy może korzystać z któregoś z wymienionych parametrów:
![image](https://github.com/KNR-PW/discord-bot/assets/117805804/4fcd4bd4-8231-4ad4-a9ad-4976544c3f8e)


https://discordpy.readthedocs.io/en/stable/api.html#discord.Permissions

pktiuk replied:
Osoba mogąca przeglądać logi serwera lub osoba, która może zarządzać rolami.

PiotrWeppo replied:
https://github.com/KNR-PW/discord-bot/pull/19

pktiuk commented on bot.py:
```diff
@@ -35,17 +38,91 @@ async def on_ready(self):
+    await ctx.reply(str(error), ephemeral=True)
+
+
+@tasks.loop(seconds=15)
+async def auto_update(
+    last_message: discord.message.Message, embed: discord.Embed, ctx: commands.Context
+):
+    """
+    Periodically updates the last sent embed. Looks for changes in embed description.
+
+    Loads the last message sent and its description.
+    Re-converts the text of its description to match the current state,
+    finally updates the message.
+
+    Args:
+        last_message (`discord.message.Message`): last sent message by bot,
+        created from the `EmbedCreator`.
+        embed (`discord.Embed`): An object from the `Discord.Embed` class that
+        will be used as the main embed.
+        ctx (discord.ext.commands.context.Context): necessary parameter when
+        accesing discoFrd server data; used by discord.ext.commands.
+    """
+    embed_description = read_from_config("embed_description")
+    output_string = convert_string(ctx, embed_description)
+    embed.description = output_string
+    now = datetime.datetime.now()
+    embed.set_footer(
+        text=f"""Last auto update: {now.strftime('%d.%m.%Y - %H:%M:%S')}"""
+    )
+    await last_message.edit(embed=embed)
```
![image](https://github.com/KNR-PW/discord-bot/assets/45544416/ce912591-052c-4b77-a9dd-698b3cd0e89f)
U mnie aktualizacja nie za bardzo działa, dodałem sobie rolę TestRola i wiadomość się nie zmieniła.

PiotrWeppo replied:
Muszę znać więcej szczegółów. Czy rolę dodałeś w polu `Edit Embed Message`, czyli zmianie głównego opisu wiadomości, czy w nowo dodanym polu? Wydaje mi się z tego co widzę, że w tym drugim. Aktualnie automatycznie aktualizowany jest tylko główny opis wiadomości, ponieważ śledzenie tekstu w każdym polu (których może być do 25) wiązałoby się w teorii ze znacznym powiększeniu ilości parametrów w pliku config i pamięci programu.

pktiuk replied:
WIeczorem postaram się to przeklikać jeszcze raz i dokładnie ci opiszę jak to zrobiłem

pktiuk replied:
0. Stworzyłem jeszcze jedną rolę `Rola2` i ją sobie przypisałem
1. Stworzyłem nowy embed /create_new
2. WYbrałem Add field
3. Wpisałem `R2 {count_members Rola2} CR2 {member Rola2}   TR {member TestRola}  CTR {count_members TestRola} `
4. Kliknąłem Ok -  już teraz w podglądzie coś jest nie tak, funkcje liczące działają dobrze, ale wypisujące ludzi nie
5. Zatwierdziłem embed
6. Odebrałem sobie wszystkie role 
7. Nadal nic się nie zmienia w wiadomości

![image](https://github.com/KNR-PW/discord-bot/assets/45544416/469fd9d4-12d6-4438-b1a0-c6102f97ea8f)

pktiuk replied:
Jako listę osób pokazuje `[None]`, mimo, że przeliczenie daje 1

PiotrWeppo replied:
dwie rzeczy są nie tak:

1. Aktualizacja śledzi jedynie zmiany w głównym opisie, tak jak pisałem wyżej, nie w dodatkowych polach.
2. Funkcja {member ...} nie służy do szukania ról, tylko do "wyplucia" imienia osoby wraz z linkiem do niej. W README masz opisane jak z niej korzystać.

Z tego co widzę wszystko działa prawidłowo.
Pierwsza funkcja zwróciła ilość 1, drugiej i trzeciej źle użyłeś a czwarta znowu zwróciła ilość 1 zgodnie z prawdą

pktiuk replied:
Ad 1. W takim razie warto byłoby to dopisać jeszcze w komendzie /help
Ad 2. Masz rację, jakieś zaciemnienie mnie złapało.

pktiuk reviewed (changes requested):
Bardzo fajnie to działa i ogólnie ma ręce i nogi.
Dziala też dość stabilnie, nie udało mi się go scrashować

PiotrWeppo committed:
Updated .gitignore to include all .ini files.
Moved all functions related to configparser to the config_parser.py

PiotrWeppo committed:
small bugfix to add_embed function.

pktiuk reviewed (approved):
Wszystko jest pięknie.

Jeśli ten branch jest gotowy do mergowania to pamiętaj zmienić jego status z drafta na zwykły.

pktiuk commented:
Jeszcze jedna, drobna poprawka, którą możesz teraz wrzucić:

```py
async def on_submit(self, interaction: Interaction[ClientT], /)
```
Taka powinna być definicja funkcji z linijki 472, aby pylint nie pokazywał jednego z błędów

PiotrWeppo committed:
Bot dynamically updates up to 5 additional fields in a message now.
Fixes so that Reset and Cancel buttons correctly work with config functions.
Several small bug fixes.

PiotrWeppo committed:
Pylint bug fix.

PiotrWeppo commented:
> Jeszcze jedna, drobna poprawka, którą możesz teraz wrzucić:
> 
> ```python
> async def on_submit(self, interaction: Interaction[ClientT], /)
> ```
> 
> Taka powinna być definicja funkcji z linijki 472, aby pylint nie pokazywał jednego z błędów

Zrobiłem tak, z tym, że "interaction: Interaction[Bot]". Natomiast zastanawia mnie ta część: ", /)". Mógłbyś wytłumaczyć co to oznacza?

PiotrWeppo commented:
Jeżeli będziesz robił review, usuń wcześniej stary plik config.ini, żeby program stworzył zaktualizowany nowy.

pktiuk commented:
> > Jeszcze jedna, drobna poprawka, którą możesz teraz wrzucić:
> > ```python
> > async def on_submit(self, interaction: Interaction[ClientT], /)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Taka powinna być definicja funkcji z linijki 472, aby pylint nie pokazywał jednego z błędów
> 
> Zrobiłem tak, z tym, że "interaction: Interaction[Bot]". Natomiast zastanawia mnie ta część: ", /)". Mógłbyś wytłumaczyć co to oznacza?

Ogólnie to kiedy tworzysz klasę potomną i nadpisujesz jedną z dziedziczonych funkcji swoją implementacją to argumenty tej funkcji powinny byc dokładnie takie same.

A jeśli chodzi o zastosowanie ukośnika jako argumentu to [słuzy on do oznaczania listy argumentów pozycyjnych](https://stackoverflow.com/questions/24735311/what-does-the-slash-mean-when-help-is-listing-method-signatures) czyli argumentów, których nie mozna podać za pomocą keywordów `funkcja(wartość1=43)`

pktiuk reviewed (approved):
Wszystko działa jak trzeba co do założeń kodu nie ma zastrzeżeń. Można mergować.

W kodzie jest trochę miejsca na poprawki, ale to się zrobi przy okazji sprzątania.
